src/core/scope.ml

```diff
@@ -69,15 +56,33 @@ type mode =
       scoped term. *)
   | M_Patt
   (** Scoping mode for patterns in the rewrite tactic. *)
-  | M_LHS  of bool * m_lhs_data
-  (** Scoping mode for rewriting rule left-hand sides. The constructor carries
-      a flag that is set to [true] if {!constructor:Terms.expo.Privat} symbols
-      are allowed, and also additional data. *)
-  | M_RHS  of bool * (string, tevar) Hashtbl.t
-  (** Scoping mode for rewriting rule righ-hand sides. The constructor carries
-      a flag that is set to [true] if {!constructor:Terms.expo.Privat} symbols
-      are allowed, and the environment for variables that we known to be bound
-      in the RHS. *)
+  | M_LHS  of
+      { m_lhs_p       : bool
+      (** True if {!constructor:Terms.expo.Privat} symbols are allowed. *)
+      ; m_lhs_indices : (string, int   ) Hashtbl.t
+      (** Stores index reserved for a pattern variable of the given name. *)
+      ; m_lhs_arities : (int   , int   ) Hashtbl.t
+      (** Stores the arity of the pattern variable at the given index. *)
+      ; m_lhs_names   : (int   , string) Hashtbl.t
+      (** Stores the name of the pattern variable at given index (if any). *)
+      ; mutable m_lhs_size : int
+      (** Stores the current known size of the environment of the RHS. *)
+      ; m_lhs_in_env  : string list
+      (** Pattern variables definitely needed in the RHS environment. *) }
+  (** Scoping mode for rewriting rule left-hand sides. *)
+  | M_RHS  of
+      { m_rhs_p             : bool
+      (** True if {!constructor:Terms.expo.Privat} symbols are allowed. *)
+      ; m_rhs_data          : (string, tevar) Hashtbl.t
+      (** Environment for variables that we know to be bound in the RHS. *)
+      ; mutable m_rhs_xvi   : int
+      (** Number of the next variable not in the LHS scoped. Incremented each
+          time such a variable is scoped. *)
+      ; mutable m_rhs_xvars : tevar list
```

Comment:
Do we have the invariant List.length m_rhs_xvars = m_rhs_xvi? If so, it would be good to mention it and rename m_rhs_xvi into m_rhs_xvars_nb.

Reply:
We rather have `m_rhs_xvi = length m_rhs_xvars + length m_lhs_vars`

Reply:
I see. It would be good to mention this in a comment. Then, m_rhs_xvi is not a number but an index and would be more appropriately named m_rhs_xvars_ind. I know that this is rather long but this will help understanding and maintenance by other people later.

Reply:
Or it is the total number of variables and could be named `m_rhs_vars_nb`, which might be even a bit clearer (since we don't invoke some index in some array)

Reply:
See my proposition, and tell me if that seems clear